### ISSUE TITLE
proposal: `prefer_const_collection_literals`

### ISSUE THREAD

**Issue by asashour** [dart-lang/linter#4029]
## prefer_const_collection_literals

**PREFER** using **const** for instantiating collection literals when possible.

Collection literals which are not assigned to a variable, are preferred to be `const`.

## Description

PREFER using 'const' for instantiating collection literals when possible.

## Details

It is preferred to use 'const' for instantiating collection literals when they are not assigned to a variable, and all the items are literals or have constant values.

## Bad Examples

```dart
  for (var i in [1, 2, 3]) {
    print(i);
  }

  if (!['cat', 'dog'].contains(value)) {
    print(value);
  }
```

## Good Examples 

```dart
  for (var i in const [1, 2, 3]) {
    print(i);
  }

  if (!const ['cat', 'dog'].contains(value)) {
    print(value);
  }

```

## Discussion

The main usage is List, but Set and Maps are also possible candidates.

### Discussion checklist

- [ ] List any existing rules this proposal modifies, complements, overlaps or conflicts with.
- [ ] List any relevant issues (reported here, the [SDK Tracker], or elsewhere).
- [ ] If there's any prior art (e.g., in other linters), please add references here.
- [ ] If this proposal corresponds to [Effective Dart] or [Flutter Style Guide] advice, please call it out. (If there isn't any corresponding advice, should there be?)
- [ ] If this proposal is motivated by real-world examples, please provide as many details as you can.  Demonstrating potential impact is _especially valuable_.

[Writing Lints]: https://github.com/dart-lang/linter/blob/main/doc/writing-lints.md
[Effective Dart]: https://dart.dev/guides/language/effective-dart
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[SDK Tracker]: https://github.com/dart-lang/sdk/issues

**Comment by srawlins** [comment]
What if it is passed as a function argument? Or given to a record literal?

What if `.add` or another modifying-method is called on it?

What if it is returned or yielded?

**Comment by bwilkerson** [comment]
I assume that most of those would be covered by the definition of "assigned", but the third question is the key question: how would the rule know which methods are non-modifying? I don't think we want a hard coded list, and there's no way to tell via static analysis (without a whole-world assumption that we can't make). I think this is the determining factor for how many false positives this rule would have.

**Comment by srawlins** [comment]
Why not a hard-coded list?

**Comment by bwilkerson** [comment]
A hard-coded list can't include any extension methods that users write, so we'd either allow modifying methods (false negative) or disallow non-modifying methods (false positive).

**Comment by srawlins** [comment]
Also, I think it would be hard/impossible to catch modifications of child elements, e.g.

```dart
void main() {
  const a = const [[1, 2, 3]];
  a[0].add(4); // runtime error
}
```

**Comment by asashour** [comment]
This is an assignment, which is to be ignored by the lint. Let's make it simple, the lint is reported only when the collection is immediately consumed with `for in` or a known non-modifying method is immediately invoked).

**Comment by srawlins** [comment]
Oh right, haha. My bad.

```dart
void main() {
  for (var e in const [[1, 2, 3]]) {
    e.add(4); // runtime error
  }
}
```

**Comment by lrhn** [comment]
User methods called normally on a collection literal can probably be assumed to be non-mutating. Mutating the collection only makes sense if you return it, and if you do that, you should have used a cascade instead.

The exception would be a method which captures the collection and returns another value containing the collection somehow. Like `List.cast`, only probably more useful, and it will want to later modify the collection, otherwise it's not a problem. I can't come up with a good example, though. It's a weird pattern.
